samcli/lib/build/bundler.py

```diff
@@ -0,0 +1,141 @@
+"""
+Handles bundler properties as needed to modify the build process
+"""
+import logging
+from copy import deepcopy
+from typing import Dict, Optional
+
+from samcli.lib.providers.provider import Stack
+from samcli.lib.providers.sam_function_provider import SamFunctionProvider
+
+LOG = logging.getLogger(__name__)
+
+ESBUILD_PROPERTY = "esbuild"
+
+
+class EsbuildBundlerManager:
+    def __init__(self, stack: Stack, template: Optional[Dict] = None):
+        self._stack = stack
+        self._previous_template = template or dict()
+
+    def esbuild_configured(self) -> bool:
+        """
+        Checks if esbuild is configured on any resource in a given stack
+        :return: True if there is a function instance using esbuild as the build method
+        """
+        function_provider = SamFunctionProvider([self._stack], ignore_code_extraction_warnings=True)
+        functions = list(function_provider.get_all())
+        for function in functions:
+            if function.metadata and function.metadata.get("BuildMethod", "") == ESBUILD_PROPERTY:
+                return True
+        return False
+
+    def enable_source_maps(self) -> Dict:
+        """
+        Appends ``NODE_OPTIONS: --enable-source-maps``, if Sourcemap is set to true
+        and sets Sourcemap to true if ``NODE_OPTIONS: --enable-source-maps`` is provided.
+        :return: Dict containing deep-copied, updated template
+        """
+        using_source_maps = False
+        invalid_node_option = False
+
+        template = deepcopy(self._previous_template)
+        resources = template.get("Resources", {})
+
+        for name, resource in resources.items():
+            metadata = resource.get("Metadata", {})
+
+            node_option_set = self._is_node_option_set(resource)
```

Comment:
Good call-out.

Reply:
The other issue with this particular use-case is that it needs to be pre-processed and not post. Otherwise the flag won't be set to true in time. I'll have to adjust this too.